Sources/MongoSwift/BSON/BSONEncoder.swift

```diff
@@ -4,19 +4,70 @@ import mongoc
 /// `BSONEncoder` facilitates the encoding of `Encodable` values into BSON.
 public class BSONEncoder {
 
+    /// Enum representing the various strategies for encoding `Date`s.
+    public enum DateEncodingStrategy {
+        /// Encode the `Date` by deferring to its default encoding implementation.
+        case deferToDate
+
+        /// Encode the `Date` as a BSON datetime object (default).
+        case bsonDate
+
+        /// Encode the `Date` as a 64-bit integer counting the number of milliseconds since January 1, 1970.
+        case millisecondsSince1970
+
+        /// Encode the `Date` as a BSON double counting the number of seconds since January 1, 1970.
+        case secondsSince1970
+
+        /// Encode the `Date` as an ISO-8601-formatted string (in RFC 339 format).
+        @available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *)
+        case iso8601
+
+        /// Encode the `Date` as a string formatted by the given formatter.
+        case formatted(DateFormatter)
+
+        /// Encode the `Date` as the value returned by the provided closure.
+        case custom((Date) throws -> BSONValue)
+    }
+
+    /// Enum representing the various strategies for encoding `UUID`s.
+    public enum UUIDEncodingStrategy {
+        /// Encode the `UUID` by deferring to its default encoding implementation.
+        case deferToUUID
```

Comment:
Is it worth discussing if this should be the actual default? What would swift users expect to be working with? @valeriomazzeo maybe you have thoughts here?

Reply:
We've chosen `.millisecondsSinceEpoch` as the default date encoding/decoding strategy for BSON spec adherence, and Swift users definitely do not expect their dates to be stored as such (all `Date`s in Swift use `Double`s for whatever reason). So we have precedent for setting it to this default.

Either way, I think we should be consistent with our defaults. If we change this one to deferred, we should also change our date encoding/decoding strategy to deferred as well.

Reply:
I suppose most people (including myself), would expect that to be encoded as a `String`, simply because it's not obvious why it should be encoded as `Binary` unless reading the documentation and I doubt that many will know about it.

I believe it goes down to your preference whether you'd like to enforce best practices as per Mongo official documentation by default, or not.

In my specific case, I believe I am going to stick with `String` initially and then optimise it later as I have some data that would need to be migrated first.

Reply:
We've decided that we're going to stick with BSON defaults. I added a bit to the docstring explaining that we're doing this with a link to the spec itself, so hopefully that give users some guidance.

On a slightly related note, I also refactored `.bsonDate` to `.bsonDateTime` to match what I did in the decoder.